radar_control/radar_control.py

```diff
@@ -228,315 +285,188 @@ def radar():
-        else:
-            print "No CP - keep polling"
+    status = RadarStatus()
+
+    new_experiment_flag = False
 
+    while not new_experiment_flag:  # poll the socket at the start until we have an experiment to run.
+        new_experiment_flag, experiment = search_for_experiment(experiment_socket, status)
 
-    while True: # REVIEW #35 This is > 250 lines, should be refactored into smaller chunks
+    new_experiment_flag = False
 
-        cpos=prog.cpo_list
-        updated_cp_received = False
+    while True:
+        # This loops through all scans in an experiment, or restarts this loop if a new experiment occurs.
 
         # Wavetables are currently None for sine waves, instead just
-        #   use a sampling freq in rads/sample
-        wavetable_dict={}
-        for cpo in range(prog.cponum):
-            wavetable_dict[cpo]=get_wavetables(prog.cpo_list[cpo]['wavetype']) # REVIEW #6 #33 Is this not needed anymore or is there a TODO to implement type of wave somewhere? We noticed it wasn't used in this file and it is a local dictionary variable
+        #   use a sampling freq in rads/sample.
+        # TODO move this to experiment handler before experiment is transferred ?
+        for slice_id, expslice in experiment.slice_dict.items():
+            # print("Slice ID {}, slice {}".format(slice_id, expslice))
+            expslice['iwavetable'], expslice['qwavetable'] = \
+                get_wavetables(expslice['wavetype'])
 
         # Iterate through Scans, AveragingPeriods, Sequences, Pulses.
-        for scan in prog.scan_objects:
-            if updated_cp_received == True:
+        for scan in experiment.scan_objects:
+            if new_experiment_flag:  # start anew on first scan if we have a new experiment. Question - should this be how we handle a new experiment that's received at the start of a new integration period - end scan after that inttime and start the new experiment?
+                try:
+                    experiment = new_experiment
+                except NameError:
+                    # new_experiment does not exist, should never happen as flag only gets set when
+                    # there is a new experiment.
+                    errmsg = 'Experiment could not be found'
+                    raise ExperimentException(errmsg)
+                new_experiment_flag = False
                 break
-            beam_remaining=True
+            beam_remaining = True  # started a new scan, therefore this must be True.
+
             # Make iterator for cycling through beam numbers
-            scan_iter=0 # REVIEW #26 Seems more apt to name this beam_iter.
-            scans_done=0
-            while (beam_remaining and not updated_cp_received):
+            aveperiods_done_list = []
+            beam_iter = 0
+            while beam_remaining and not new_experiment_flag:
```

Comment:
This is much better than before. It's still a very important loop, with a lot of different data structures used.
A series of comments throughout the loop breaking it up into a high level overview of what is happening would be good to add.
Also, since the loop is larger than one page, adding a comment on the last line of the for loop (seq_num_start += nave) to indicate that it's the end of that loop would help with readability.